### PR TITLE
Accelerate CI parallelizing over OMP configs

### PR DESCRIPTION
**Pull request by vincentmr**
### Before submitting

Please complete the following checklist when submitting a PR:

- [x] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      [`tests`](../tests) directory!

- [x] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [x] Ensure that the test suite passes, by running `make test`.

- [x] Add a new entry to the `.github/CHANGELOG.md` file, summarizing the
      change, and including a link back to the PR.

- [x] Ensure that code is properly formatted by running `make format`. 

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**
`tests_linux_cpp::cpptests` take ~9-10 minutes each. There are 3 `enable_kernel_omp`/`enable_kernel_avx_stream` options that need be tested, which takes ~30 minutes.

**Description of the Change:**
Use a matrix to parallelize builds over 3 runners. Update a few path filters and workflows with `determine_runner` dependency.

**Benefits:**

**Possible Drawbacks:**

**Related GitHub Issues:**
[sc-62648]

### CODE REVIEW

File: .github/CHANGELOG.md [hunk 1]
```diff
@@ -18,6 +21,7 @@
 This release contains contributions from (in alphabetical order):
 
 Amintor Dusko
+Vincent Michaud-Rioux
```
Comment (AmintorDusko) [inline .github/CHANGELOG.md:24]:
```suggestion
Amintor Dusko, Vincent Michaud-Rioux
```

**Review by AmintorDusko — APPROVED** [review]
Looks good to me! Could you please provide an estimation of the new workflow run times?

**Comment by vincentmr** [comment]
> Looks good to me! Could you please provide an estimation of the new workflow run times?

Yes, it is written in the `Context` section of the description.

**Comment by AmintorDusko** [comment]
> > Looks good to me! Could you please provide an estimation of the new workflow run times?
> 
> Yes, it is written in the `Context` section of the description.

Thanks! :pray:

**Pull request merged**